### ISSUE TITLE
How to give frame into the Yolov5_StrongSORT_OSNet sequentially?

### ISSUE THREAD

**Issue by tgrmev** [mikel-brostrom/Yolov5_StrongSORT_OSNet#601]
### Search before asking

- [X] I have searched the Yolov5_StrongSORT_OSNet [issues](https://github.com/mikel-brostrom/Yolov5_StrongSORT_OSNet/issues) and found no similar bug report.

### Question

Hello, I am trying to  do object tracking on the video that I subscribe with ROS2.  First of all, I only use yolov5 for detecting and load the model with `model = torch.hub.load("./yolov5", 'custom', path="./yolov5/yolov5m.pt", source='local') # local repo`. I needed to crop the each detected object and that is why, I am using your repo. Normally, after loading yolov5 model, I gave each frame to the model as you can see in the code snippet. My question is how can I load SORT model inside the ROS2 node and how can I give the each frame into SORT?
`

    def listener_callback(self, data):

      current_frame = self.br.compressed_imgmsg_to_cv2(data)
      processed_image = self.model(current_frame)    
      processed_image.save()
      cv2.imshow("camera", processed_image.ims[0])`

**Comment by mikel-brostrom** [comment]
I am not that familiar with ROS2. But you would need to move the whole code into a node and read from a stream instead of individual images

**Comment by tgrmev** [comment]
Actually, I asked my question with ROS2, cause I wanted to explain what I am doing in detail. We can basically say that how can I load the SORT model from a python file. I added, how I call yolov5 from a python file(FIRST CODE). I could call your repo with subprocess(SECOND CODE) but I can not use subprocess in my work. How can I call your repo in a py file not using subprocess?

```
#####FIRST CODE#####

import torch
import torch.nn as nn
import os
import cv2 


model = torch.hub.load('ultralytics/yolov5', 'yolov5m') 
print("custom yolov5 loaded")

img_fdr="./yolov5/horse.png"

##show the detected image
processed_image = model(img_fdr)

processed_image.save(save_dir='./yolov5/runs/detect') 

cv2.imshow("camera", processed_image.ims[0])

detected_obj=processed_image.crop(save_dir='./yolov5/runs/detect')

cv2.waitKey(0)
cv2.destroyAllWindows() 

```

For the second code, I would like to save your model somehow like first code.


```
#####SECOND CODE#####
print(os.getcwd())
os.chdir(r"./Yolov5_StrongSORT_OSNet")
print(os.getcwd())

img_fdr="./video.mp4"

subprocess.run([f"python3","track.py", "--source", img_fdr])


cv2.waitKey(0)
cv2.destroyAllWindows() 

```
I also found a similar questions for the another repository. It could help to understand the my question better. [https://github.com/nwojke/deep_sort/issues/252](url)

**Comment by mikel-brostrom** [comment]
StrongSORT is not only a model there is a lot of surrounding code for the association process.
You can however create a class

https://github.com/mikel-brostrom/Yolov5_StrongSORT_OSNet/blob/eab08d1a7bf4b3c63251e9400fe8fe6df6a1c977/trackers/multi_tracker_zoo.py#L13

then you pass the detections:

```python
outputs = strongsort.update(det.cpu(), im0)
```

and finally you extract the tracking results

```python
for j, (output, conf) in outputs
    bboxes = output[0:4]
    id = output[4]
    cls = output[5]
```

**Comment by tgrmev** [comment]
It works. Thanks a lot @mikel-brostrom

**Comment by 0xNOY** [comment]
Sorry for the very elementary question, but what does the variable `det` and its method `cpu()` mean in the following code?
I would appreciate it if you could answer. Thank you in advance.
> ``` python
> outputs = strongsort.update(det.cpu(), im0)
> ```

**Comment by mikel-brostrom** [comment]
det = detections
Applying `.cpu()` to them implies moving the detections gpu to cpu

**Comment by tgrmev** [comment]
@0xNOY . For my case, I got an error when I write `outputs = strongsort.update(det.cpu(), im0).` In that case, I changed init function in detect.py file to `self.feature = np.asarray(feature.cpu().detach().numpy(), dtype=np.float32).` Additionally, det function comes from yoloModel. You can call it with `det=processed_image.xyxy[0]` as a tensor.

**Comment by mikel-brostrom** [comment]
The CI pipeline shows that all works as expected, no changes has to be made. If you have problems, start for a clean slate, just re-clone and run again
